libbpf-tools/bashreadline.c

```diff
@@ -0,0 +1,218 @@
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_size)
+{
+	struct str_t *e = data;
+	struct tm *tm;
+	char ts[16];
+	time_t t;
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%m:%S", tm);
+
+	printf("%-9s %-7d %s\n", ts, e->pid, e->str);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warn("lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+static char *find_readline_so()
+{
+	const char *bash_path = "/bin/bash";
+	FILE *fp;
+	char *line = NULL;
+	char *sep, *path_start, *sp;
+	size_t line_sz;
+	off_t func_off;
+	char *result = NULL;
+
+	func_off = get_elf_func_offset(bash_path, "readline");
+	if (func_off >= 0)
+		return strdup(bash_path);
+
+	/*
+	 * Try to find libreadline.so if the symbol of readline is not
+	 * defined in bash
+	 */
+	fp = popen("ldd /bin/bash", "r");
+	if (fp == NULL)
+		goto cleanup;
+	while (getline(&line, &line_sz, fp) >= 0) {
+		if (strstr(line, "libreadline.so") == NULL)
+			continue;
+		sep = strstr(line, " => ");
+		if (sep == NULL)
+			goto cleanup;
+		path_start = sep + 4;
+		while (isspace(*path_start))
+			path_start++;
+		sp = strstr(path_start, " ");
+		if (sp == NULL)
+			goto cleanup;
+		path_start[sp - path_start] = 0;
+		if (strstr(path_start, "/libreadline.so") == NULL)
+			goto cleanup;
+		result = strdup(path_start);
+		break;
+	}
+
+cleanup:
+	if (fp)
+		fclose(fp);
+	if (line)
+		free(line);
+	return result;
+}
+
+static void sig_int(int signo)
+{
+	exiting = 1;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct bashreadline_bpf *obj = NULL;
+	struct perf_buffer_opts pb_opts;
+	struct perf_buffer *pb = NULL;
+	char *readline_so_path;
+	off_t func_off;
+	int err;
+	int fd;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	if (libreadline_path) {
+		readline_so_path = libreadline_path;
+	} else if ((readline_so_path = find_readline_so()) == NULL) {
+		warn("failed to find readline\n");
+		return 1;
+	}
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		warn("failed to increase rlimit: %d\n", err);
+		goto cleanup;
+	}
+
+	obj = bashreadline_bpf__open_and_load();
+	if (!obj) {
+		warn("failed to open and load BPF object\n");
+		goto cleanup;
+	}
+
+	func_off = get_elf_func_offset(readline_so_path, "readline");
+	if (func_off < 0) {
+		warn("cound not find readline in %s\n", readline_so_path);
+		goto cleanup;
+	}
+
+	obj->links.printret = bpf_program__attach_uprobe(obj->progs.printret, true, -1,
+							 readline_so_path, func_off);
+	err = libbpf_get_error(obj->links.printret);
+	if (err) {
+		warn("failed to attach readline: %d\n", err);
+		goto cleanup;
+	}
+
+	fd = bpf_map__fd(obj->maps.events);
+
+	pb_opts.sample_cb = handle_event;
+	pb_opts.lost_cb = handle_lost_events;
+	pb = perf_buffer__new(fd, PERF_BUFFER_PAGES, &pb_opts);
```

Comment:
nit: you could have just inlined bpf_map__fd(obj->maps.events) here, it would be fine